### PR TITLE
Patch #499: Refactor find_host_config to lib.find_module_in_config

### PR DESCRIPTION
**Pull request by BigRoy**
This PR is related to #499 

This now makes sure the way the `config.{host}` module is retrieved per host matches across each integration using `avalon.lib.find_module_in_config` as a replacement for the `find_host_config` function that each `avalon.{host}` module implemented.

Aside of that, I've made the variable names consistently `config_host` as opposed to replacing the `config` variable that could previously had left some confusion thinking it was the config module, but was actually the `config.{host}` mdoule.

And of course, as per #499 this should now allow `config.{host}` module to not exist at all for each of the integrations with only a logged warning along the lines of: `Config has no '{config}.{host}' module.`

Additionally this makes these changes:

- Add `uninstall()` function to `avalon.fusion` as that was missing.
- Match Houdini and Fusion metadata collecting to Maya implementation, which had an optimization.

### Discussion

**NOTE:** This code should still be tested. However I wanted to set up this PR to discuss the code choices.

1. Is the code clear enough?

2. This imports the `find_module_in_host` function from `avalon.lib` as it became more readable than doing the following. However I believe this was frowned upon by the code style, but I couldn't find that mentioned anywhere.
```python
# in this PR
from ..lib import find_module_in_host
find_module_in_host(config, "maya")

# the other option
import ..lib as avalon_lib
avalon_lib.find_module_in_host(config, "maya")
```
    
3. Can we simplify further? Could the function be simplified to: `get_submodule`? This would shorten the function name and make it more generic. Then however, it logging the warning message feels to "specific" to the use cases. Maybe removing the warning is actually fine?

### CODE REVIEW

**Comment by BigRoy** [comment]
Actually, this could be done much easier. 

We should remove the code from the individual host integrations and move it to the global install function. Will update the PR.

**Comment by BigRoy** [comment]
With https://github.com/getavalon/core/pull/501/commits/3f1aa64bc8e53cde982f09f67a213a13c2e03408 the Host Integrations now don't need to manually call `install()` and `uninstall()` on the configuration.

The `{config}.{host}.install()` and `{config}.{host}.uninstall()` are called from respectively `avalon.pipeline.install()` and `avalon.pipeline.uninstall()` now when the studio configuration has these methods exposed.

File: avalon/lib.py [hunk 1]
```diff
@@ -281,3 +281,23 @@ def modules_from_path(path):
...
+        module (str): The config's submodule to search.
+    
+    Returns:
+        types.ModuleType or None: The module, if found.
+        
+    """
+    config_name = config.__name__
+    module_name = "%s.%s" % (config_name, module)
+    try:
+        return importlib.import_module(module_name)
```
Comment (davidlatwe) [inline avalon/lib.py]: Houdini and Nuke raised `NameError: global name 'importlib' is not defined` at this line.

File: avalon/lib.py [hunk 2]
```diff
@@ -281,3 +281,23 @@ def modules_from_path(path):
         modules.append(module)
 
     return modules
+    
+
```
Comment (davidlatwe) [inline avalon/lib.py]: A few places pop PEP8 warning due to blank line containing white space.

**Comment by davidlatwe** [comment]
> However I believe this was frowned upon by the code style, but I couldn't find that mentioned anywhere.

I think it's here, in [CONTRIBUTING # Code Quality - Architecture](https://github.com/getavalon/core/blob/master/CONTRIBUTING.md#code-quality---architecture) at row A8.

> Could the function be simplified to: get_submodule?

Yeah, we can simplify that, but I lean to name it `find_submodule`, since I think the word `get` sounds like "You will get what you want for sure", and `find` is a bit uncertain which fit the scenario "You may not found the submodule you want, be careful".

> Then however, it logging the warning message feels to "specific" to the use cases. Maybe removing the warning is actually fine?

I more like to keep the message, and rephrase it to like:
```python
log_.warning("Could not find '%s' in module '%s'." % (submodule, module))
```
And the variables need to be renamed as well.

**Comment by BigRoy** [comment]
Thanks @davidlatwe for the thorough check. Admittedly I wrote the previous code at home and it was untested (as mentioned in the original post) because I was on holiday leave. Back in the office Today, cleaned up the code, listened to the lovely hound that is back to his regular woof-woof (thanks Marcus!) and tested the integrations in Maya, Houdini and Fusion to be sure. All good now.

Can you double check the code style and let me know if there's still anything out of the ordinary?

File: avalon/lib.py [hunk 3]
```diff
@@ -281,3 +282,23 @@ def modules_from_path(path):
...
+        types.ModuleType or None: The module, if found.
+
+    """
+    name = "%s.%s" % (module.__name__, submodule)
+    try:
+        return importlib.import_module(name)
+    except ImportError as exc:
+        if str(exc) != "No module name {}".format(name):
+            log_.warning("Could not find '%s' in module: %s" % (submodule,
+                                                                module))
```
Comment (jasperges) [inline avalon/lib.py]: Just a minor thing: the preferred method for string formatting for logging functions is to pass arguments (see [here](https://docs.python.org/2/library/logging.html?#logging.Logger.debug) for more info). So I thinks it's better to write:

```python
log_.warning(
    "Could not find '%s' in module: %s",
    submodule,
    module,
)

File: avalon/houdini/pipeline.py [hunk 4]
```diff
@@ -10,8 +10,8 @@
 
 # Local libraries
 from . import lib
-from ..lib import logger
-from avalon import api, schema
+from ..lib import logger, find_submodule
+from avalon import api
```
Comment (jasperges) [inline avalon/houdini/pipeline.py]: Shouldn't this be `from .. import api`?

Comment (BigRoy) [reply avalon/houdini/pipeline.py]: It sure should. I just removed the `schema` part but will fix it right up! Well spotted.

File: avalon/lib.py (hunk 3 above)
Comment (BigRoy) [reply avalon/lib.py]: That's entirely true. I'll reformat hit. However I believe in many places the code is written with actual string formatting... which we should actually update too. Usually the difference is really only worth it for code that needs to be heavily optimized (runs very frequently) because the `logging` formatting in the function will only trigger when an actual log handler is connected to it and the message actually needs to be formatted. Which in our case usually is the case anyway.

File: avalon/lib.py [hunk 5]
```diff
@@ -281,3 +282,23 @@ def modules_from_path(path):
...
+
+    Args:
+        module (types.ModuleType): The module to search in.
+        submodule (str): The submodule name to find.
+
+    Returns:
+        types.ModuleType or None: The module, if found.
+
+    """
+    name = "%s.%s" % (module.__name__, submodule)
```
Comment (jasperges) [inline avalon/lib.py]: Maybe also use `.format()` here to be consistent?

File: avalon/houdini/pipeline.py (hunk 4 above)
Comment (BigRoy) [reply avalon/houdini/pipeline.py]: Resolved with https://github.com/getavalon/core/pull/501/commits/986157351ac5bd6ea8dd5d22cb1ec4ac892952c3

File: avalon/lib.py (hunk 3 above)
Comment (BigRoy) [reply avalon/lib.py]: Resolved with https://github.com/getavalon/core/pull/501/commits/6d894d13133e7726d9e76bf64fec1c11985732a9

File: avalon/lib.py (hunk 5 above)
Comment (BigRoy) [reply avalon/lib.py]: Resolved with: https://github.com/getavalon/core/pull/501/commits/5aea369f9d3818d89c81297a1e32764d34180c5f

**Comment by BigRoy** [comment]
Just found [this odd exception printing](https://github.com/getavalon/core/pull/501/files#diff-95efa4b4467abab66473e907fd11d122R278-R280) which is not a change by this PR but has been there for quite some time. Is there any Python version where that actually prints as a formatted string? I don't believe so. 

**Python 2**
```python
err = "bar"
print("Skipped: \"%s\" (%s)", mod_name, err)
# ('Skipped: "%s" (%s)', 'foo', 'bar')
```

**Python 3**
```python
err = "bar"
print("Skipped: \"%s\" (%s)", mod_name, err)
# Skipped: "%s" (%s) foo bar
```

I'll patch that up too.

**Comment by BigRoy** [comment]
[This](https://github.com/getavalon/core/pull/501#issuecomment-570183899) is fixed with: https://github.com/getavalon/core/pull/501/commits/241e05d83e53ad128dbb3038dab493122cb3fea9

Retested in Maya + Houdini to be sure. All seems to work as expected. 👍

**Comment by jasperges** [comment]
@BigRoy Did a quick check with Blender with the changes you mentioned and everything seems to work fine.

**Comment by BigRoy** [comment]
Merging this.
